### PR TITLE
Implement gRPC for peer connection

### PR DESCRIPTION
**Pull request by willgreenberg**
- Adds gRPC functionality as an alternative to HTTP for peer communication + tests
- Moves HTTP into its own subpackage (and puts gRPC into a subpackage)

### CODE REVIEW

File: grpc.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+func (gp *GRPCFetchProtocol) NewFetcher(address string) (RemoteFetcher, error) {
+	conn, err := grpc.Dial(address, gp.PeerDialOptions...)
+	if err != nil {
+		return nil, err
+	}
+	return &grpcFetcher{address: address, conn: conn}, nil
+}
+
+// GRPCPeerServer implements the PeerServer interface generated by the Peer pb service
+type GRPCPeerServer struct {
```
Comment (dfinkel) [inline grpc.go]: I'd separate the GRPC server-side handler into its own file. It shares very little code with the client-side and there's no point in mixing the methods up like this.

File: grpc.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	parentUniverse *Universe
```
Comment (dfinkel) [inline grpc.go]: Let's rename `parentUniverse` to `universe`, there's no other universe this will be interacting with.

File: grpc.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+// GRPCFetchProtocol specifies GRPC specific options for GRPC-based peer communcation
+type GRPCFetchProtocol struct {
+	// connection set up configurations for all peers
+	PeerDialOptions []grpc.DialOption
+}
+
+type grpcFetcher struct {
+	address string
+	conn    *grpc.ClientConn
```
Comment (dfinkel) [inline grpc.go]: I'd hold onto the `PeerClient` in this struct, so you don't have to construct a new one every time `Fetch` is called.

File: grpc.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+// Fetch here implements the RemoteFetcher interface for sending Gets to peers over an RPC connection
+func (g *grpcFetcher) Fetch(ctx context.Context, galaxy string, key string) ([]byte, error) {
+	client := pb.NewPeerClient(g.conn)
+	resp, err := client.GetFromPeer(ctx, &pb.GetRequest{
+		Galaxy: galaxy,
+		Key:    key})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to GET [%s]: %v", galaxy, err)
```
Comment (dfinkel) [inline grpc.go]: We're not issuing an HTTP GET request here, let's make the error message clearer about this being a GRPC request to the `GetFromPeer` method.

File: grpc.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	return &grpcFetcher{address: address, conn: conn}, nil
+}
+
+// GRPCPeerServer implements the PeerServer interface generated by the Peer pb service
+type GRPCPeerServer struct {
+	parentUniverse *Universe
+}
+
+// RegisterGRPCServer creates a new GRPC server and GRPCHandler and registers them for RPC use;
```
Comment (sachinagada) [inline grpc.go]: `RegisterGRPCServer` isn't creating a new GRPC server, it's registering `GRPCPeerServer` with the grpc server that's passed in

File: grpc.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+// the purpose of the GRPCHandler is to keep track of the parent universe in order to access the groups for Get calls
```
Comment (sachinagada) [inline grpc.go]: update comment, `GRPCHandler` no longer exists and remove "parent" from `parent universe`

File: grpc.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+func RegisterGRPCServer(universe *Universe, grpcServer *grpc.Server) {
+	pb.RegisterPeerServer(grpcServer, &GRPCPeerServer{parentUniverse: universe})
+}
+
+// GetFromPeer implements the generated pb.GalaxycacheServer interface, making an internal Get() after receiving a remote call from a peer
```
Comment (sachinagada) [inline grpc.go]:
```suggestion
// GetFromPeer implements the generated PeerServer interface, making an internal Get() after receiving a remote call from a peer
```

File: grpc.go (hunk 4 above)
Comment (sachinagada) [reply grpc.go]: It would be helpful to know which peer we failed to get info from. We can log that from where Fetch is being called from because that passes in the peer we should go to

File: grpcclient.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+type grpcFetcher struct {
+	address string
+	conn    *grpc.ClientConn
+	client  pb.GalaxyCacheClient
+}
+
+// NewGRPCFetchProtocol creates a GRPC fetch protocol to be
+// passed into a Universe constructor
```
Comment (dfinkel) [inline grpcclient.go]: Let's drop the tail of this docstring. "to be passed into a Universe constructor" isn't descriptive of what this does, and it may not be true if something wraps a `Universe` instance.

It would be better to provide an [example](https://golang.org/pkg/testing/#hdr-Examples) and explicit documentation about how to wire everything together in the `README`.

I'd probably also reword this to say something like:

 > `NewGRPCFetchProtocol` creates a fetch-protocol implementation using GRPC for communicating with peers. Users without TLS certificates on the peers operating as servers should specify `grpc.WithInsecure()` as one of the arguments.

File: grpcclient.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
+/*
+ Copyright 2019 Will Greenberg
```
Comment (dfinkel) [inline grpcclient.go]: Copyright Vimeo Inc.

File: grpcclient.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+ Adapted from https://github.com/charithe/gcgrpcpool/blob/master/gcgrpcpool.go
```
Comment (dfinkel) [inline grpcclient.go]: Drop this reference. There isn't enough of that left to be worth crediting.

File: grpcclient.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// Fetch here implements the RemoteFetcher interface for
+// sending Gets to peers over an RPC connection
+func (g *grpcFetcher) Fetch(ctx context.Context, galaxy string, key string) ([]byte, error) {
+	resp, err := g.client.GetFromPeer(ctx, &pb.GetRequest{
+		Galaxy: galaxy,
+		Key:    key})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to fetch from peer over RPC [%s, %s]: %v", galaxy, g.address, err)
```
Comment (dfinkel) [inline grpcclient.go]: Let's quote `galaxy` and `g.address`; I also generally use `%s` for errors since the `fmt` package's `printf`-like functions know how to call the `Error()` method of `error`s:
```suggestion
		return nil, fmt.Errorf("Failed to fetch from peer over RPC [%q, %q]: %s", galaxy, g.address, err)
```

File: grpc_test.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+		}
+		if suffix := ":" + key; !strings.HasSuffix(string(value), suffix) {
+			t.Errorf("Get(%q) = %q, want value ending in %q", key, value, suffix)
+		}
+		t.Logf("Get key=%q, value=%q (peer:key)", key, value)
+	}
+	grpcServer.GracefulStop()
+}
+
+func makeGRPCServerUniverse(ctx context.Context, t testing.TB, addresses []string, listener net.Listener) {
```
Comment (dfinkel) [inline grpc_test.go]: This is misnamed. It doesn't just create the GRPC server, it also runs it and shuts it down when its done.
It would be better to call this `runTestPeerGRPCServer`

File: grpc_test.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	}
+
+	universe := NewUniverse(NewGRPCFetchProtocol(grpc.WithInsecure()), "shouldBeIgnored")
+	defer func() {
+		shutdownErr := universe.Shutdown()
+		if shutdownErr != nil {
+			t.Errorf("Error on shutdown: %s", shutdownErr)
+		}
+	}()
+	grpcServer := grpc.NewServer()
```
Comment (dfinkel) [inline grpc_test.go]: We're not running this RPC server at all. Let's just drop it and all its references including the `GracefulShutdown` below.

File: grpc_test.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for _, listener := range peerListeners {
+		go makeGRPCServerUniverse(ctx, t, peerAddresses, listener)
+	}
+
+	for _, key := range testKeys(nGets) {
+		var value StringCodec
+		if err := g.Get(dummyCtx, key, &value); err != nil {
```
Comment (dfinkel) [inline grpc_test.go]: let's delete the `dummyCtx`, the `ctx` above is perfectly good.

File: grpc_test.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+	const (
+		nRoutines = 5
+		nGets     = 100
+	)
+
+	var peerAddresses []string
+	var peerListeners []net.Listener
+
+	for i := 0; i < nRoutines; i++ {
+		newListener := pickFreeAddr(t)
```
Comment (dfinkel) [inline grpc_test.go]: This really isn't gaining you much, just call `net.Listen` directly here.

File: grpc/grpc_test.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+		t.Errorf("Error setting peers: %s", err)
+	}
+
+	getter := gc.GetterFunc(func(ctx context.Context, key string, dest gc.Codec) error {
+		dest.UnmarshalBinary([]byte(":" + key))
+		return nil
+	})
+	universe.NewGalaxy("peerFetchTest", 1<<20, getter)
+	go func() {
+		defer listener.Close()
```
Comment (sachinagada) [inline grpc/grpc_test.go]: There's an error that `listener.Close()` returns that we aren't handling right now with this defer statement. You can use a `defer func()` similar to what you do on line 94 and do error handling inside the function

File: grpc/grpc_test.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+	for _, listener := range peerListeners {
+		go runTestPeerGRPCServer(ctx, t, peerAddresses, listener)
+	}
+
+	for _, key := range testKeys(nGets) {
+		var value gc.StringCodec
+		if err := g.Get(ctx, key, &value); err != nil {
+			t.Fatal(err)
+		}
+		if suffix := ":" + key; !strings.HasSuffix(string(value), suffix) {
```
Comment (sachinagada) [inline grpc/grpc_test.go]: It looks like the value will always be equal to `:key` (ex: `:0`, `:1`) so we don't have to check for suffix. We can directly check for `":"+key == string(value)`

File: grpc/grpc_test.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+	g := universe.NewGalaxy("peerFetchTest", 1<<20, getter)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for _, listener := range peerListeners {
+		go runTestPeerGRPCServer(ctx, t, peerAddresses, listener)
+	}
+
+	for _, key := range testKeys(nGets) {
```
Comment (sachinagada) [inline grpc/grpc_test.go]: I don't think testKeys is a necessary function. We can range through nGets and create the key inside the for loop with `strconv.Itoa(i)`

File: http/http_test.go [hunk 19]
```diff
@@ -73,7 +78,7 @@ func TestHTTPHandler(t *testing.T) {
 	}
 
 	for _, key := range testKeys(nGets) {
```
Comment (sachinagada) [inline http/http_test.go]: Same thing about testKeys not being a necessary function.

File: grpc/grpcclient.go [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+	conn    *grpc.ClientConn
+	client  pb.GalaxyCacheClient
+}
+
+// NewGRPCFetchProtocol creates a fetch-protocol implementation
+// using GRPC for communicating with peers. Users without TLS
+// certificates on the peers operating as servers should specify
+// grpc.WithInsecure() as one of the arguments.
+func NewGRPCFetchProtocol(dialOpts ...grpc.DialOption) *GRPCFetchProtocol {
+	return &GRPCFetchProtocol{PeerDialOptions: dialOpts}
```
Comment (dfinkel) [inline grpc/grpcclient.go]: let's append the `ocgrpc` instrumentation option here:
https://godoc.org/go.opencensus.io/plugin/ocgrpc#example-ClientHandler

This should take care of propagating trace contexts and request metrics, which people can opt into by registering the appropriate views.

File: grpc/grpcserver.go [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+	pb.RegisterGalaxyCacheServer(grpcServer, &GRPCGalaxyCacheServer{universe: universe})
+}
+
+// GetFromPeer implements the generated GalaxyCacheServer
+// interface, making an internal Get() after receiving a
+// remote call from a peer
+func (gp *GRPCGalaxyCacheServer) GetFromPeer(ctx context.Context, req *pb.GetRequest) (*pb.GetResponse, error) {
+	galaxy := gp.universe.GetGalaxy(req.Galaxy)
+	if galaxy == nil {
+		return nil, fmt.Errorf("Unable to find galaxy [%s]", req.Galaxy)
```
Comment (dfinkel) [inline grpc/grpcserver.go]: Let's set an appropriate status code here: [`codes.InvalidArgument`](https://godoc.org/google.golang.org/grpc/codes#InvalidArgument)
```suggestion
		return nil, status.Errorf(codes.InvalidArgument, "Unable to find galaxy [%s]", req.Galaxy)
```

(see below for more detail about `status`)

File: http/http.go [hunk 22]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package galaxycache
+package httppeerconn
```
Comment (dfinkel) [inline http/http.go]: The package name should match the directory in most cases.

Let's call this package `http`. (It has both the peer connection and the http handler, so it's misleading to call it `httppeerconn`.)

File: grpc/grpc_test.go (hunk 18 above)
Comment (willgreenberg) [reply grpc/grpc_test.go]: Chatted with David and Sergio about this -- it's a good point, but stylistically for the sake of this test (and keeping it Go-y with the ranged for loop) we're thinking we'll leave this one as is

File: http/http_test.go (hunk 19 above)
Comment (willgreenberg) [reply http/http_test.go]: Same as grpc_test.go:77

File: grpc/grpc_test.go (hunk 16 above)
Comment (willgreenberg) [reply grpc/grpc_test.go]: This caused a panic because it led to possible logging occurring after the test had completed -- reverted for now to still not handle the error and added a TODO for resolving it

File: grpc/grpcclient.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+	address string
+	conn    *grpc.ClientConn
+	client  pb.GalaxyCacheClient
+}
+
+// NewGRPCFetchProtocol creates a fetch-protocol implementation
+// using GRPC for communicating with peers. Users without TLS
+// certificates on the peers operating as servers should specify
+// grpc.WithInsecure() as one of the arguments.
+func NewGRPCFetchProtocol(dialOpts ...grpc.DialOption) *GRPCFetchProtocol {
```
Comment (sachinagada) [inline grpc/grpcclient.go]: I see that we added open census tracing/metrics for http. Let's do that for GRPC as well

Comment (willgreenberg) [reply grpc/grpcclient.go]: I did that right there in grpcclient.go, no? Appended an ocgrpc-wrapped dial opt.

Comment (sachinagada) [reply grpc/grpcclient.go]: You did. Do we want to call any methods on the wrapped handler like [TagRPC](https://godoc.org/go.opencensus.io/plugin/ocgrpc#ClientHandler.TagRPC)

Comment (sergiosalvatore) [reply grpc/grpcclient.go]: Hmm, possibly -- but as this PR has gotten a bit large I think we can come come back with a metrics-focused PR.

Comment (dfinkel) [reply grpc/grpcclient.go]: This is fine as-is.

We shouldn't need to call `HandleRPC`, that's something that should be handled automatically by the actual owner of the `GRPCServer` registering `ocgrpc.ServerHandler` as a `grpc.StatsHandler` at [Server creation][1].

[1]: https://godoc.org/go.opencensus.io/plugin/ocgrpc#example-ServerHandler

File: grpc/grpcclient.go [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+	client := pb.NewGalaxyCacheClient(conn)
+	return &grpcFetcher{address: address, conn: conn, client: client}, nil
+}
+
+// Fetch here implements the RemoteFetcher interface for
+// sending Gets to peers over an RPC connection
+func (g *grpcFetcher) Fetch(ctx context.Context, galaxy string, key string) ([]byte, error) {
+	resp, err := g.client.GetFromPeer(ctx, &pb.GetRequest{
+		Galaxy: galaxy,
+		Key:    key})
```
Comment (dfinkel) [inline grpc/grpcclient.go]: nit: I tend not to close struct literals on the same line as the last entry, instead, I tend to leave a trailing comma and close them on the next line. (it makes adding new fields later much cleaner)

```suggestion
		Key:    key,
	})
```

File: grpc/grpcserver.go [hunk 25]
```diff
@@ -0,0 +1,61 @@
+/*
+ Copyright 2019 Vimeo Inc.
+ Adapted from https://github.com/charithe/gcgrpcpool/blob/master/gcgrpcpool.go
```
Comment (dfinkel) [inline grpc/grpcserver.go]: There's not much left from this other implementation that isn't a trivial GRPC client/service example. Let's drop all mentions of it.
